File: libs/sdk-bindings/bindings-android/lib/src/main/kotlin/breez_sdk_notification/BreezSdkConnector.kt
package breez_sdk_notification

import breez_sdk.BlockingBreezServices
import breez_sdk.ConnectRequest
import breez_sdk.EventListener
import breez_sdk.LogEntry
import breez_sdk.LogStream
import breez_sdk.connect

class BreezSdkConnector {
    companion object {
        private const val TAG = "BreezSdkConnector"

        private var breezSDK: BlockingBreezServices? = null

        internal fun connectSDK(
            connectRequest: ConnectRequest,
            sdkListener: EventListener,
            logger: LogStream?
        ): BlockingBreezServices {
            synchronized(this) {
                if (breezSDK == null) {
                    logger?.log(
                        LogEntry(TAG, "Connecting to Breez SDK", "DEBUG")
                    )
                    breezSDK = connect(connectRequest, sdkListener)
                    logger?.log(LogEntry(TAG, "Connected to Breez SDK", "DEBUG"))
                } else logger?.log(LogEntry(TAG, "Already connected to Breez SDK", "DEBUG"))

                return breezSDK!!
            }
        }
    }
}

File: libs/sdk-bindings/bindings-swift/Sources/BreezSDKNotification/BreezSDKConnector.swift
import Foundation
import os.log

#if DEBUG && true
fileprivate var logger = OSLog(
    subsystem: Bundle.main.bundleIdentifier!,
    category: "BreezSDKConnector"
)
#else
fileprivate var logger = OSLog.disabled
#endif

class BreezSDKConnector {
    private static var breezSDK: BlockingBreezServices? = nil
    fileprivate static var queue = DispatchQueue(label: "BreezSDKConnector")
    fileprivate static var sdkListener: EventListener? = nil
    
    static func register(connectRequest: ConnectRequest, listener: EventListener) throws -> BlockingBreezServices? {
        try BreezSDKConnector.queue.sync { [] in
            BreezSDKConnector.sdkListener = listener
            if BreezSDKConnector.breezSDK == nil {
                BreezSDKConnector.breezSDK = try BreezSDKConnector.connectSDK(connectRequest: connectRequest)
            }
            return BreezSDKConnector.breezSDK
        }
    }
    
    static func unregister() {
        BreezSDKConnector.queue.sync { [] in
            BreezSDKConnector.sdkListener = nil
        }
    }
    
    static func connectSDK(connectRequest: ConnectRequest) throws -> BlockingBreezServices? {
        // Connect to the Breez SDK make it ready for use
        os_log("Connecting to Breez SDK", log: logger, type: .debug)
        let breezSDK = try connect(req: connectRequest, listener: BreezSDKEventListener())
        os_log("Connected to Breez SDK", log: logger, type: .debug)
        return breezSDK
    }
}

class BreezSDKEventListener: EventListener {
    func onEvent(e: BreezEvent) {
        BreezSDKConnector.queue.async { [] in
            BreezSDKConnector.sdkListener?.onEvent(e: e)
        }
    }
}
